zcash_primitives/src/transaction/mod.rs

```diff
@@ -53,6 +53,85 @@ impl fmt::Display for TxId {
     }
 }
 
+/// The set of defined transaction format versions.
+///
+/// This is serialized in the first four or eight bytes of the transaction format, and
+/// represents valid combinations of the `(overwintered, version, version_group_id)`
+/// transaction fields.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum TxVersion {
+    Sprout(u32),
+    Overwinter,
+    Sapling,
+    #[cfg(feature = "zfuture")]
+    ZFuture,
+}
+
+impl TxVersion {
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let header = reader.read_u32::<LittleEndian>()?;
+        let overwintered = (header >> 31) == 1;
+        let version = header & 0x7FFFFFFF;
+
+        if overwintered {
+            match (version, reader.read_u32::<LittleEndian>()?) {
+                (OVERWINTER_TX_VERSION, OVERWINTER_VERSION_GROUP_ID) => Ok(TxVersion::Overwinter),
+                (SAPLING_TX_VERSION, SAPLING_VERSION_GROUP_ID) => Ok(TxVersion::Sapling),
+                #[cfg(feature = "zfuture")]
+                (ZFUTURE_TX_VERSION, ZFUTURE_VERSION_GROUP_ID) => Ok(TxVersion::ZFuture),
+                _ => Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "Unknown transaction format",
+                )),
+            }
+        } else {
+            Ok(TxVersion::Sprout(version))
+        }
+    }
+
+    pub fn header(&self) -> u32 {
+        match self {
+            TxVersion::Sprout(v) => *v,
+            TxVersion::Overwinter => OVERWINTER_TX_VERSION | 1 << 31,
+            TxVersion::Sapling => SAPLING_TX_VERSION | 1 << 31,
+            #[cfg(feature = "zfuture")]
+            TxVersion::ZFuture => ZFUTURE_TX_VERSION | 1 << 31,
+        }
```

Comment:
Possibly more resilient to future changes (since we never want the overwintered bit to be unset for future versions):
```suggestion
        let overwintered = match self {
            TxVersion::Sprout(v) => 0,
            _ => 1 << 31,
        };
        match self {
            TxVersion::Sprout(v) => *v,
            TxVersion::Overwinter => OVERWINTER_TX_VERSION,
            TxVersion::Sapling => SAPLING_TX_VERSION,
            #[cfg(feature = "zfuture")]
            TxVersion::ZFuture => ZFUTURE_TX_VERSION,
        } | overwintered
```